### PR TITLE
Add mouse support to the Input Manager

### PR DESCRIPTION
**Pull request by paked**

### CODE REVIEW

File: axis.go [hunk 1]
```diff
@@ -39,3 +39,47 @@ func (keys AxisKeyPair) Value() float32 {
 
 	return AxisNeutral
 }
+
+const (
+	// The vertical mouse axis
```
Comment (otraore) [inline axis.go]: Follow [godoc style](https://blog.golang.org/godoc-documenting-go-code) specifically: 

> Notice this comment is a complete sentence that begins with the name of the element it describes.

File: axis.go [hunk 2]
```diff
@@ -39,3 +39,47 @@ func (keys AxisKeyPair) Value() float32 {
...
+	AxisMouseVert = 0
+	// The horizontal mouse axis
```
Comment (otraore) [inline axis.go]: Same as comment above

File: axis.go [hunk 3]
```diff
@@ -39,3 +39,47 @@ func (keys AxisKeyPair) Value() float32 {
...
+	AxisMouseHori = 1
+)
+
+// AxisMouse is an axis of the mouse direction.
```
Comment (otraore) [inline axis.go]: Grammar? Doesn't sound right.

File: axis.go [hunk 4]
```diff
@@ -39,3 +39,47 @@ func (keys AxisKeyPair) Value() float32 {
...
+type AxisMouse struct {
+	direction int
+
+	old float32
+}
+
+// NewAxisMouse creates a new Mouse Axis in either direction AxisMouseVert or AxisMouseHori.
+func NewAxisMouse(d int) *AxisMouse {
```
Comment (otraore) [inline axis.go]: Suggestion: Rename from AxisMouse -> MousAxis

File: axis.go [hunk 5]
```diff
@@ -39,3 +39,47 @@ func (keys AxisKeyPair) Value() float32 {
...
+	AxisMouseVert = 0
+	// The horizontal mouse axis
+	AxisMouseHori = 1
+)
+
+// AxisMouse is an axis of the mouse direction.
+type AxisMouse struct {
+	direction int
+
+	old float32
```
Comment (otraore) [inline axis.go]: Internal documentation for those working or browsing through engo, not clear at first what this does.

File: axis.go [hunk 6]
```diff
@@ -39,3 +39,47 @@ func (keys AxisKeyPair) Value() float32 {
...
+	// The vertical mouse axis
+	AxisMouseVert = 0
+	// The horizontal mouse axis
+	AxisMouseHori = 1
+)
+
+// AxisMouse is an axis of the mouse direction.
+type AxisMouse struct {
+	direction int
+
```
Comment (otraore) [inline axis.go]: Unneeded space?

File: axis.go [hunk 7]
```diff
@@ -39,3 +39,47 @@ func (keys AxisKeyPair) Value() float32 {
...
+		old:       old,
+	}
+}
+
+// Value returns the delta of a mouse movement.
+func (am *AxisMouse) Value() float32 {
+	var diff float32
+
+	if am.direction == AxisMouseHori {
+		diff = Input.Mouse.X - am.old
```
Comment (otraore) [inline axis.go]: not sure how I feel about this space

File: axis.go [hunk 8]
```diff
@@ -39,3 +39,47 @@ func (keys AxisKeyPair) Value() float32 {
...
+
+		am.old = Input.Mouse.X
+	} else {
+		diff = Input.Mouse.Y - am.old
+
```
Comment (otraore) [inline axis.go]: or this one

File: demos/mouse-follow/follow.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import (
+	"image/color"
+
+	"engo.io/ecs"
+	"engo.io/engo"
+)
+
+type DefaultScene struct{}
+
+type Guy struct {
```
Comment (otraore) [inline demos/mouse-follow/follow.go]: Suggestion for later PRs: Move the "Guy" object to demoutils, it's being re-used a bunch.

File: demos/mouse-follow/follow.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+	ecs.BasicEntity
+	engo.RenderComponent
+	engo.SpaceComponent
+}
+
+func (*DefaultScene) Preload() {
+	// This could be done individually: engo.Files.Add("data/icon.png"), etc
+	// Second value (false) says whether to check recursively or not
+	engo.Files.AddFromDir("data", false)
```
Comment (otraore) [inline demos/mouse-follow/follow.go]: Suggest we just use 
```go
engo.Files.Add("data/icon.png") 
```
considering it's just one file

File: demos/mouse-follow/follow.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+func (s *FollowSystem) Update(dt float32) {
+	for _, e := range s.entities {
+		e.SpaceComponent.Position.X += engo.Input.Axis(engo.DefaultMouseXAxis).Value()
+		e.SpaceComponent.Position.Y += engo.Input.Axis(engo.DefaultMouseYAxis).Value()
+	}
+}
+
+func main() {
+	opts := engo.RunOptions{
+		Title:          "Scale Demo",
```
Comment (otraore) [inline demos/mouse-follow/follow.go]: Wrong title for this demo

File: input.go [hunk 12]
```diff
@@ -17,6 +17,8 @@ func NewInputManager() *InputManager {
 
 // InputManager contains information about all forms of input.
 type InputManager struct {
+	Mouse Mouse
```
Comment (otraore) [inline input.go]: Starting to sound like a broken record but, DOCUMENTASHUNNNNNN

File: mouse.go [hunk 13]
```diff
@@ -7,6 +7,14 @@ import (
 
 const MouseSystemPriority = 100
 
+type Mouse struct {
```
Comment (otraore) [inline mouse.go]: DOCUMENTASHUNNN

File: mouse.go [hunk 14]
```diff
@@ -7,6 +7,14 @@ import (
...
+	X, Y             float32
```
Comment (otraore) [inline mouse.go]: DOCUMENTASHUNNN

File: demos/mouse-follow/follow.go (hunk 9 above)
Comment (mewmew) [inline demos/mouse-follow/follow.go]: Same comment as in
https://github.com/EngoEngine/ecs/pull/18#discussion_r60993404
> Lets make it gender neutral. `Player`, `Hero`, `Character` or something along those lines.

File: axis.go (hunk 8 above)
Comment (paked) [inline axis.go]: This space was accidental.

File: axis.go (hunk 4 above)
Comment (paked) [inline axis.go]: > 22:08 @otraore The reason why AxisMouse is called AxisMouse is to a) Make auto correct work nicely, b) keep in sync with the other Axis* structs

Response from gitter.

**Comment by paked** [comment]
Removed in progress because this is ready for a final review/merge @EtienneBruines @otraore.

File: axis.go [hunk 15]
```diff
@@ -39,3 +39,48 @@ func (keys AxisKeyPair) Value() float32 {
...
+	// AxisMouseHori is vertical mouse axis
+	AxisMouseHori = 1
+)
+
+// AxisMouse is an axis for a single x or y component of the Mouse. The value returned from it is
+// the delta movement, since the previous call and it is not constrained by the AxisMin and AxisMax values.
+type AxisMouse struct {
+	// direction is the value storing either AxisMouseVert and AxisMouseHori. It determines which directional
+	// component to operate on.
+	direction int
```
Comment (EtienneBruines) [inline axis.go]: Let's not make this an `int`, but creat something like `type AxisMouseDirection uint`?  Same for lines 45 and 47 above (they would be of that new type)

File: axis.go [hunk 16]
```diff
@@ -40,27 +40,29 @@ func (keys AxisKeyPair) Value() float32 {
 	return AxisNeutral
 }
 
+type AxisMouseDirection uint
```
Comment (EtienneBruines) [inline axis.go]: Not sure if we want to document this.

**Comment by EtienneBruines** [comment]
LGTM. Thank you for your work :smile:

**Pull request merged**